dist/secure-handlebars.js

```diff
@@ -10141,5 +10055,699 @@ module.exports.create = overrideHbsCreate;
+    else if ((chr >= '\x01'   && chr <= '\x08') || 
+             (chr >= '\x0E'   && chr <= '\x1F') ||
+             (chr >= '\x7F'   && chr <= '\x9F') ||
+             (chr >= '\uFDD0' && chr <= '\uFDEF') ||
+             chr === '\x0B' || chr === '\uFFFE' || chr === '\uFFFF') {
+        input[i] = '\uFFFD';
+    } 
+    // U+1FFFE, U+1FFFF, U+2FFFE, U+2FFFF, U+3FFFE, U+3FFFF, 
+    // U+4FFFE, U+4FFFF, U+5FFFE, U+5FFFF, U+6FFFE, U+6FFFF, 
+    // U+7FFFE, U+7FFFF, U+8FFFE, U+8FFFF, U+9FFFE, U+9FFFF, 
+    // U+AFFFE, U+AFFFF, U+BFFFE, U+BFFFF, U+CFFFE, U+CFFFF, 
+    // U+DFFFE, U+DFFFF, U+EFFFE, U+EFFFF, U+FFFFE, U+FFFFF, 
+    // U+10FFFE, and U+10FFFF 
+    else if ((nextChr === '\uDFFE' || nextChr === '\uDFFF') && 
+             (  chr === '\uD83F' || chr === '\uD87F' || chr === '\uD8BF' || chr === '\uD8FF' || 
+                chr === '\uD93F' || chr === '\uD97F' || chr === '\uD9BF' || chr === '\uD9FF' || 
+                chr === '\uDA3F' || chr === '\uDA7F' || chr === '\uDABF' || chr === '\uDAFF' || 
+                chr === '\uDB3F' || chr === '\uDB7F' || chr === '\uDBBF' || chr === '\uDBFF')) {
+        input[i] = input[i+1] = '\uFFFD';
+    }
+}
+
+function ConvertBogusCommentToComment(i) {
+    // convert !--. i.e., from <* to <!--*
+    this.input.splice(i, 0, '!', '-', '-'); 
+    this.inputLen += 3;
+
+    // convert the next > to -->
+    this.on('preCanonicalize', PreCanonicalizeConvertBogusCommentEndTag);
+}
+
+function PreCanonicalizeConvertBogusCommentEndTag(state, i, endsWithEOF) {
+    if (this.input[i] === '>') {
+        // remove itself from the listener list
+        this.off('preCanonicalize', PreCanonicalizeConvertBogusCommentEndTag);
+
+        // convert [>] to [-]->
+        this.input.splice(i, 0, '-', '-');
+        this.inputLen += 2;
+
+        this.emit('bogusCommentCoverted', state, i, endsWithEOF);
+    }
+}
+
+// those doctype states (52-67) are initially treated as bogus comment state, but they are further converted to comment state 
+// Canonicalize() will create no more bogus comment state except the fake (context-parser treats <!doctype as bogus) one hardcoded as <!doctype html> that has no NULL inside
+var statesRequiringNullReplacement = [
+//    0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+/*0*/ 0, 0, 0, 1, 0, 1, 1, 1, 0, 0,
+/*1*/ 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+/*2*/ 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,
+/*3*/ 1, 1, 0, 0, 1, 1, 1, 1, 1, 1,
+/*4*/ 1, 0, 0, 0, 1, 0, 1, 1, 1, 1,
+/*5*/ 1, 1
+];
+// \uFFFD replacement is not required by the spec for DATA state
+statesRequiringNullReplacement[htmlState.STATE_DATA] = 1;
+
+function Canonicalize(state, i, endsWithEOF) {
+    
+    this.emit('preCanonicalize', state, i, endsWithEOF);
+
+    var reCanonicalizeNeeded = true,
+        input = this.input,
+        chr = input[i], nextChr = input[i+1],
+        potentialState = this._getNextState(state, i, endsWithEOF),
+        nextPotentialState = this._getNextState(potentialState, i + 1, endsWithEOF);
```

Comment:
nextChr = input[i + 1] will pick up undefined
_getNextState() have done i >= length check, if so, returns -1

the subsequent logic only read from these values for checks. would be fine